crypto/params.c

```diff
@@ -0,0 +1,365 @@
+        CASE(uint64_t);
+        }
+        break;
+    }
+    return 0;    
+#undef CASE
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string(const char *key, char *buf,
+                                            size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING, buf, bsize,
+                                rsize);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string_ptr(const char *key, char **buf,
+                                                size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING_PTR, buf, bsize,
+                                rsize);
+}
+
+int OSSL_PARAM_get_utf8_string(const OSSL_PARAM *p, char **val, size_t max_len)
+{
+    char *q;
+    int is_ptr;
+
+    if (val == NULL || p == NULL)
+        return 0;
+    is_ptr = p->data_type != OSSL_PARAM_UTF8_STRING_PTR;
+    if (!is_ptr && p->data_type != OSSL_PARAM_UTF8_STRING)
+        return 0;
+
+    q = *val;
+    if (q == NULL) {
+        if (is_ptr) {
+            *val = *(char **)p->buffer;
+            return 1;
+        }
+        if ((q = OPENSSL_malloc(p->buffer_size)) == NULL)
+            return 0;
+        *val = q;
+        memcpy(q, p->buffer, p->buffer_size);
+        return 1;
+    }
+    if (max_len < p->buffer_size)
+        return 0;
+    if (is_ptr)
+        memcpy(q, *(char **)p->buffer, p->buffer_size);
+    else
+        memcpy(q, p->buffer, p->buffer_size);
+    return 1;
+}
+
+int OSSL_PARAM_set_utf8_string(OSSL_PARAM *p, char *val)
+{
+    if (p == NULL)
+        return 0;
+
+    if (val != NULL) {
+        const size_t len = strlen(val) + 1;
+
+        SET_RETURN_SIZE(p, len);
+        switch (p->data_type) {
+        case OSSL_PARAM_UTF8_STRING:
+            if (p->buffer_size >= len) {
+                memcpy(p->buffer, val, len);
+                return 1;
+            }
+            break;
+
+        case OSSL_PARAM_UTF8_STRING_PTR:
+            *(char **)p->buffer = val;
+            return 1;
+        }
+    } else
+        SET_RETURN_SIZE(p, 0);
+    return 0;
+}
+
+OSSL_PARAM OSSL_PARAM_construct_octet_string(const char *key, void *buf,
+                                             size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_OCTET_STRING, buf, bsize,
+                                rsize);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_octet_string_ptr(const char *key, void **buf,
+                                                 size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_OCTET_STRING_PTR, buf, bsize,
+                                rsize);
+}
+
+int OSSL_PARAM_get_octet_string(const OSSL_PARAM *p, void **val, size_t max_len,
+                                size_t *used)
+{
+    char *q;
+    int is_ptr;
+
+    if (val == NULL || p == NULL)
+        return 0;
+    if (used != NULL)
+        *used = p->return_size != NULL && *p->return_size != 0
+                ? *p->return_size : p->buffer_size;
```

Comment:
This came from my own confusion.  Only `p->buffer_size` should be used here.  See https://github.com/openssl/openssl/pull/8400#issuecomment-470004598 for an explanation

Reply:
Fixed.